### PR TITLE
add support for implicit TLS

### PR DESCRIPTION
**Pull request by drakkan**
This is not a standard and should not be used anymore but there are still clients requiring it.

I reused the `TLSRequired` setting, if you prefer I can add a different setting for this mode, thank you

### CODE REVIEW

**Comment by fclairamb** [comment]
I'm not a huge fan of this setting. What do you think about at least defining a "sort of enum" for it ?

```golang
type TLSRequirement int

const (
    ClearOrEncrypted TLSRequirement = iota
    MandatoryEncryption
    ImplicitEncryption
)
```

**Comment by drakkan** [comment]
> I'm not a huge fan of this setting. What do you think about at least defining a "sort of enum" for it ?
> 
> ```go
> type TLSRequirement int
> 
> const (
>     ClearOrEncrypted TLSRequirement = iota
>     MandatoryEncryption
>     ImplicitEncryption
> )
> ```

I agree, I would like to add `NoEncryption` to the enum too (disable PROT and AUTH), I'll send a patch later today, thank you!

**Comment by drakkan** [comment]
Hi, please take a look at this updated version. I also did some test cleanup and used [testify](https://github.com/stretchr/testify) for new tests. If you don't like testify I'll change that part, thank you!

**Comment by drakkan** [comment]
Currently we use two ftp client libraries for test cases, ideally we should use only one (github.com/secsy/goftp seems newer) and I also would like to use testify. What do you think about?

**Comment by fclairamb** [comment]
Hi @drakkan,

Sorry for the merge you had to do. I modified some things around the CI to actually take advantage of the concurrent test execution design.

I agree that having one client is better than two. And yes I'm ok with using testify. 
I'd like to have as few dependencies as possible so that people can easily integrate this lib into their project but we don't have to be as conservative for tests and testify makes the test code cleaner, so definitely worth using it.

File: client_handler_test.go [hunk 1]
```diff
@@ -17,26 +19,54 @@ func TestConcurrency(t *testing.T) {
...
+func TestTLSMethods(t *testing.T) {
+	t.Run("without-tls", func(t *testing.T) {
+		cc := clientHandler{
+			server: NewTestServer(t, true),
+		}
+		require.False(t, cc.HasTLSForControl())
+		require.False(t, cc.HasTLSForTransfers())
+	})
+
+	t.Run("wit-implicit-tls", func(t *testing.T) {
```
Comment (fclairamb) [inline client_handler_test.go]: with

**Comment by fclairamb** [comment]
And yes the `NoEncryption` enum is a good idea.

**Comment by drakkan** [comment]
> And yes the `NoEncryption` enum is a good idea.

I don't think it is required, if I want to disable encryption I can simply don't return a TLS config

**Comment by drakkan** [comment]
> Hi @drakkan,
> 
> Sorry for the merge you had to do. I modified some things around the CI to actually take advantage of the concurrent test execution design.
> 

No problem, finally you added codecov, great!

> I agree that having one client is better than two. And yes I'm ok with using testify.
> I'd like to have as few dependencies as possible so that people can easily integrate this lib into their project but we don't have to be as conservative for tests and testify makes the test code cleaner, so definitely worth using it.

**Comment by fclairamb** [comment]
Yep, codecov should make the PR submission effort easier.

**Review by fclairamb — APPROVED** [review]
Looks very good to me

**Pull request merged**